File: js-ts/D/after/device/EfDeviceRepository.ts
// Low-level module: Concrete implementation using a mock ORM (e.g., Entity Framework).
import { Device } from "./Device";
import { IDeviceRepository } from "./IDeviceRepository";

export class EfDeviceRepository implements IDeviceRepository {
    private devices: Device[] = [
        new Device(1, "Thermostat"),
        new Device(2, "Smart Light"),
        new Device(3, "Security Camera")
    ];

    public getAll(): Device[] {
        // In a real app, this would be a call to a DbContext
        return this.devices;
    }

    public getById(id: number): Device | undefined {
        // In a real app, this would be a call to a DbContext
        return this.devices.find(d => d.id === id);
    }
}

File: csharp/D/After/EfDeviceRepository.cs

namespace After;

public class EfDeviceRepository : IDeviceRepository
{
    private readonly ApiDbContext apiDbContext;

    public EfDeviceRepository(ApiDbContext apiDbContext)
    {
        this.apiDbContext = apiDbContext;
    }

    public List<Device> GetAll()
    {
        return apiDbContext.Devices.ToList();
    }

    public Device? GetById(int id)
    {
        return apiDbContext.Devices.FirstOrDefault(x => x.Id == id);
    }
}
